# Lesson 2

# Data Cleaning
Today we will discuss data cleaning. Data cleaning is an important piece of Data Science work. In most cases the data we have to work with is inconsistent, incomplete or do not have the fields we need. The process of transforming the data to get ready to gain insight from it is called data cleaning. 

This lecture will teach the key components through examples. In this tutorial we will be working with a Wine Quality dataset from Kaggle. We will continue using this dataset for a majority of the lectures so make sure you understand the different columns etc. 

## Beginner 



### Import required packages 
The first thing we need to do is istall the rquired packages. We install all the functionality from pandas with a general import because we will be using many different functions and this makes a general import cleaner than a specifying each function. 

In [1]:
import pandas as pd 
import numpy as np

### Load Data

In [60]:
wine_data = pd.read_csv('winequalityN.csv')

### Examine Data
The first thing we do after we have loaded the file is to look at it. 

In [34]:
wine_data.head(10)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   
5  white            8.1              0.28         0.40             6.9   
6  white            6.2              0.32         0.16             7.0   
7  white            7.0              0.27         0.36            20.7   
8  white            6.3              0.30         0.34             1.6   
9  white            8.1              0.22         0.43             1.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   
5      0.050                 30.0                  97.0   0.9951  3.26   
6      0.045                 30.0                 136.0   0.9949  3.18   
7      0.045                 45.0                 170.0   1.0010  3.00   
8      0.049                 14.0                 132.0   0.9940  3.30   
9      0.044                 28.0                 129.0   0.9938  3.22   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6  
5       0.44     10.1        6  
6       0.47      9.6        6  
7       0.45      8.8        6  
8       0.49      9.5        6  
9       0.45     11.0        6

Secondly we want to look ath the column headers and datatypes. 

In [4]:
print(wine_data.shape)

(6497, 13)


In [5]:
print(wine_data.columns)

Index(['type', 'fixed acidity', 'volatile acidity', 'citric acid',
       'residual sugar', 'chlorides', 'free sulfur dioxide',
       'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol',
       'quality'],
      dtype='object')


In [6]:
print(wine_data.dtypes)

type                     object
fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object


<font color='red'>Thinking question:</font> Are these the data types you would expect based on your intuitive understanding of the columns based on the names?

### Check data for completeness

Before we jump into analysis we need to know if the data is complete. In the last lecture we taklked about how python represents missing values. In python these are noted as Null or None values. Below we sum over all rows and all columns for the total number of missing values. 
  
<font color='red'>Thinking question:</font> Why do we care about missing values?

In [7]:
missing_values = wine_data.isnull().sum().sum()
total_values = wine_data.size
print('The wine dataset has {0} missing values out of {1} which corresponds to {2}%'\
      .format(missing_values, total_values,100*missing_values/total_values))



The wine dataset has 38 missing values out of 84461 which corresponds to 0.04499117936088846%


To figure out how to deal with the missing data we first need to know how they spread out accross the columns

In [8]:
wine_data.isnull().sum()

type                     0
fixed acidity           10
volatile acidity         8
citric acid              3
residual sugar           2
chlorides                2
free sulfur dioxide      0
total sulfur dioxide     0
density                  0
pH                       9
sulphates                4
alcohol                  0
quality                  0
dtype: int64

We can see that most of the missing data is in the _fixed_acidity_ column. However it can still be heard to get an intuitive sense of what this means. Let's look at one row with missing data. 

In [9]:
wine_data[wine_data.isna().any(axis=1)].head(2)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
17  white            NaN              0.66         0.48             1.2   
33  white            6.2              0.12         0.34             NaN   

    chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
17      0.029                 29.0                  75.0   0.9892  3.33   
33      0.045                 43.0                 117.0   0.9939  3.42   

    sulphates  alcohol  quality  
17       0.39     12.8        8  
33       0.51      9.0        6

Going back to our earlier findings, we know that only 0.05% of the values are missing. Before we styart removing data we want to see how many rows would be affected. The command below tells us how many rows are missing data. 

In [10]:
wine_data.apply(lambda x: sum(x.isnull().values), axis = 1).sum()

38

In [12]:
100*38/6497

0.58488533169155

### Remove missing data

This tells us that no row is missing more than 1 value, but it is still a very small fraction of the total rows so it seems safe to go ahead and clean the data by sumply removing the rows with missing data. 

In [15]:
cleaned_on_column = wine_data.dropna(axis=1)

In [16]:
cleaned_wine_data = wine_data.dropna(axis=0)

We want to make sure the new frame contains all the same information by comparing the columns and making sure we only dropped the rows containing missing values. 

In [18]:
cleaned_wine_data.isnull().sum()

type                    0
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [14]:
print('original data dimmensions',wine_data.shape)

original data dimmensions (6497, 13)


In [17]:
print('cleaned data dimmensions',cleaned_on_column.shape)
print('cleaned data dimmensions',cleaned_wine_data.shape)

cleaned data dimmensions (6497, 6)
cleaned data dimmensions (6463, 13)


## Intermediate 
### Convert Data 
Looking at the data types for each column we can see that _quality_ is an integer, _int64_ . It might make more sense to treat this as a string because we can also think of it as a label. To achieve this we can cast/convert the entire column. 

In [98]:
wine_data.astype({'quality': 'str'}, inline=True)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0     white            100             0.270         0.36           20.70   
1     white            100             0.300         0.34            1.60   
2     white            100             0.280         0.40            6.90   
3     white            100             0.230         0.32            8.50   
4     white            100             0.230         0.32            8.50   
5     white            100             0.280         0.40            6.90   
6     white            100             0.320         0.16            7.00   
7     white            100             0.270         0.36           20.70   
8     white            100             0.300         0.34            1.60   
9     white            100             0.220         0.43            1.50   
10    white            100             0.270         0.41            1.45   
11    white            100             0.230         0.40            4.20   
12    white            100             0.180         0.37            1.20   
13    white            100             0.160         0.40            1.50   
14    white            100             0.420         0.62           19.25   
15    white            100             0.170         0.38            1.50   
16    white            100             0.480         0.04            1.10   
17    white            100             0.660         0.48            1.20   
18    white            100             0.340         0.42            1.10   
19    white            100             0.310         0.14            7.50   
20    white            100             0.660         0.48            1.20   
21    white            100             0.310         0.38            2.90   
22    white            100             0.260         0.42            1.70   
23    white            100             0.670         0.14            1.50   
24    white            100             0.270         0.41            1.30   
25    white            100             0.250         0.32            9.00   
26    white            100             0.240         0.35            1.00   
27    white            100             0.280         0.39            8.70   
28    white            100             0.270         0.48            1.10   
29    white            100             0.320         0.36            2.00   
...     ...            ...               ...          ...             ...   
6467    red            100             0.510         0.14            1.90   
6468    red            100             0.360         0.53            2.20   
6469    red            100             0.380         0.14            2.20   
6470    red            100             0.690         0.32            2.20   
6471    red            100             0.580         0.20            2.40   
6472    red            100             0.310         0.78           13.90   
6473    red            100             0.520         0.40            2.20   
6474    red            100             0.300         0.63            1.60   
6475    red            100             0.700         0.15            5.10   
6476    red            100             0.670         0.15            1.80   
6477    red            100             0.560         0.09            1.70   
6478    red            100             0.350         0.33            2.40   
6479    red            100             0.560         0.09            1.70   
6480    red            100             0.715         0.10            2.60   
6481    red            100             0.460         0.29            2.10   
6482    red            100             0.320         0.44            2.40   
6483    red            100             0.390         0.44            2.60   
6484    red            100             0.310         0.41            2.40   
6485    red            100             0.610         0.11            1.80   
6486    red            100               NaN         0.33            2.50   
648

In [19]:
wine_data_converted = wine_data.astype({'quality': 'str'})
print(wine_data_converted.dtypes)

type                     object
fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                  object
dtype: object


One quick notes about dataframes in pandas. If we just run  
`wine_data.astype({'quality': 'str'})`  
the changes are not saved but rather just a copy that we cannot access later is created.  
If we run either of the two:  
- `wine_data = wine_data.astype({'quality': 'str'})` 
- `wine_data.astype({'quality': 'str'}, inplace=True)`  
The wine data variable will refer to the updated dataframe. 

### Impute missing values 
Sometimes we want to replace the missing values rather than removing the rowes that contains missing data. But itn is important that we are aware of what ch<nges this can cause. Let's take a look at the distribution of the original column excluding the missing values.  

In [20]:
wine_data[['fixed acidity']].describe()

fixed acidity
count    6487.000000
mean        7.216579
std         1.296750
min         3.800000
25%         6.400000
50%         7.000000
75%         7.700000
max        15.900000

In [21]:
wine_data[['fixed acidity']].fillna(0).describe()

fixed acidity
count    6497.000000
mean        7.205472
std         1.326280
min         0.000000
25%         6.400000
50%         7.000000
75%         7.700000
max        15.900000

By replacing all Na values with 0 the mean valye decreased slightly, 7.217 to 7.205. 

If we wanted to replace all missing values with 0 for the whole dataframe we run the command below. 

In [22]:
wine_data['fixed acidity'].fillna(wine_data['fixed acidity'].mean()).describe()

count    6497.000000
mean        7.216579
std         1.295751
min         3.800000
25%         6.400000
50%         7.000000
75%         7.700000
max        15.900000
Name: fixed acidity, dtype: float64

In [23]:
wine_data['fixed acidity'].fillna(wine_data['fixed acidity'].median()).describe()

count    6497.000000
mean        7.216246
std         1.295779
min         3.800000
25%         6.400000
50%         7.000000
75%         7.700000
max        15.900000
Name: fixed acidity, dtype: float64

In [ ]:
wine_data_removed = wine_data.fillna(0)

As we can see below the resulting dataframe as the same dimmensions as the original data but no missing values. 

In [ ]:
print('original data dimmensions',wine_data.shape)
print('cleaned data dimmensions',wine_data_removed.shape)
wine_data_removed.isnull().sum()

If we do not want to simply fill in with 0s we can impute using the mean of the column

In [ ]:
wine_data['fixed acidity'].fillna(wine_data['fixed acidity'].mean()).describe()

Deciding what technique is most suitable for imputing is more of an art of the challenge. If you have enough data that simply removing the rows will not affect the balance of your data that is the simplest. That is not always the case. For those situations when you have to impute you should pick the method that allows the underlaying distribution of the data to change as little as possible. 

## Advanced 
Let's say we want to add a column calles _rating_ depending on the quality labeling the wine as simply good or bad. We can see that there are 7 different qualities, 3,4,5,6,7,8,9. Let's say we want to label everything less than 6 as bad and more than 5 good. 

In [ ]:
wine_data['rating'] = np.where(wine_data['quality']>5, 'Good', 'Bad')
print(wine_data.head(5))

In [ ]:
print(wine_data['quality'].value_counts())
print(wine_data['rating'].value_counts())

## Conclusion
You made it all the way to the end of _Data Cleaning_ you should make sure you fully understand everything in the basic section and if you want everything in the intermediate and/or advanced as well. We will use many of these concepts going forward so it might make your future weeks a little bit easier. 

By the end of this class you should be able to:
- Determine if your data has missing values
- Remove the missing data from your dataset

Think about the follow two questions for the feedback section: 
1. What are some potential issues that comes rom simply removing rows with missing data in your dataset? 
2. Did any of the underlaying disributions change in your data when you removed the rows? if yes, what impact do you think this will have on susequent analyses. 
2. Come up with at least 1 alternative approach you could use and examine how it compares to simply removing the missing values. 

### Updates section 
#### Fill column with random numbers

In [92]:
wine_data.head(10)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   
5  white            8.1              0.28         0.40             6.9   
6  white            6.2              0.32         0.16             7.0   
7  white            7.0              0.27         0.36            20.7   
8  white            6.3              0.30         0.34             1.6   
9  white            8.1              0.22         0.43             1.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   
5      0.050                 30.0                  97.0   0.9951  3.26   
6      0.045                 30.0                 136.0   0.9949  3.18   
7      0.045                 45.0                 170.0   1.0010  3.00   
8      0.049                 14.0                 132.0   0.9940  3.30   
9      0.044                 28.0                 129.0   0.9938  3.22   

   sulphates  alcohol  quality  test  
0       0.45      8.8        6   100  
1       0.49      9.5        6   100  
2       0.44     10.1        6   100  
3       0.40      9.9        6   100  
4       0.40      9.9        6   100  
5       0.44     10.1        6   100  
6       0.47      9.6        6   100  
7       0.45      8.8        6   100  
8       0.49      9.5        6   100  
9       0.45     11.0        6   100

Replaces the missing elements in a column by random elements from that same column. 

In [96]:
from random import uniform
ceiling = wine_data['fixed acidity'].max()
floor = wine_data['fixed acidity'].min()
wine_data['fixed acidity'] = wine_data["fixed acidity"].apply(lambda x: uniform(floor, ceiling) if np.isnan(x) else 100)

#### Impute by prediction 